packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -0,0 +1,257 @@
+   * When referring to the resource in a resource policy, use this as ARN.
+   *
+   * (Depending on the resource type, this needs to be '*' in a resource policy).
+   *
+   * @default Same as regular resource ARNs
+   */
+  resourceSelfArns?: string[];
+}
+
+/**
+ * Options for a grant operation that only applies to principals
+ */
+export interface GrantOnPrincipalOptions extends CommonGrantOptions {
+  /**
+   * Construct to report warnings on in case grant could not be registered
+   */
+  scope?: cdk.IConstruct;
+}
+
+/**
+ * Options for a grant operation to both identity and resource
+ */
+export interface GrantOnPrincipalAndResourceOptions extends CommonGrantOptions {
+  /**
+   * The resource with a resource policy
+   *
+   * The statement will always be added to the resource policy.
+   */
+  resource: IResourceWithPolicy;
+
+  /**
+   * When referring to the resource in a resource policy, use this as ARN.
+   *
+   * (Depending on the resource type, this needs to be '*' in a resource policy).
+   *
+   * @default Same as regular resource ARNs
+   */
+  resourceSelfArns?: string[];
+}
+
+export class Permissions {
+  /**
+   * Grant the given permissions to the principal
+   *
+   * The permissions will be added to the principal policy primarily, falling
+   * back to the resource policy if necessary. The permissions must be granted
+   * somewhere.
+   *
+   * - Trying to grant permissions to a principal that does not admit adding to
+   *   the principal policy while not providing a resource with a resource policy
+   *   is an error.
+   * - Trying to grant permissions to an absent principal (possible in the
+   *   case of imported resources) leads to a warning being added to the
+   *   resource construct.
+   */
+  public static grantWithResource(options: GrantWithResourceOptions): Grant {
+    const result = Permissions.grantOnPrincipal({
+      ...options,
+      scope: options.resource
+    });
+
+    if (result.success) { return result; }
+
+    const statement = new PolicyStatement()
+      .addActions(...options.actions)
+      .addResources(...(options.resourceSelfArns || options.resourceArns))
+      .addPrincipal(options.principal!);
+
+    options.resource.addToResourcePolicy(statement);
+
+    return grantResult({ addedToResource: true, statement, options });
+  }
+
+  /**
+   * Try to grant the given permissions to the given principal
+   *
+   * Absence of a principal leads to a warning, but failing to add
+   * the permissions to a present principal is not an error.
+   */
+  public static grantOnPrincipal(options: GrantOnPrincipalOptions): Grant {
+    if (!options.principal) {
+      if (options.scope) {
+        // tslint:disable-next-line:max-line-length
+        options.scope.node.addWarning(`Could not add grant for '${options.actions}' on '${options.resourceArns}' because the principal was not available. Add the permissions by hand.`);
+      }
+      return grantResult({ principalMissing: true, options });
```

Comment:
I wonder if we should just return the statement here even if we didn't actually add to the policy so `statement` will always be valid.